Firestore/Example/Tests/Integration/API/FIRDatabaseTests.mm

```diff
@@ -1240,4 +1241,79 @@ - (void)testClientCallsAfterShutdownFail {
       NSException, @"The client has already been shutdown.");
 }
 
+- (void)testMaintainsPersistenceAfterRestarting {
+  FIRDocumentReference *doc = [self documentRef];
+  FIRFirestore *firestore = doc.firestore;
+  FIRApp *app = firestore.app;
+  NSString *appName = app.name;
+  FIROptions *options = app.options;
+
+  NSDictionary<NSString *, id> *initialData = @{@"foo" : @"42"};
+  [self writeDocumentRef:doc data:initialData];
+
+  // -clearPersistence() requires Firestore to be shut down. Shutdown FIRApp to emulate the way
+  // an end user would do this.
+  [self shutdownFirestore:firestore];
+  [self.firestores removeObject:firestore];
+  [self deleteApp:app];
+
+  // Reconfigure the app and assert that persistent storage persisted.
```

Comment:
Again, this comment describes *what* the code is doing without any insight into *why* it's doing so. Your intuition to make a comment here is right, you just need to comment on why--the code already says what it's doing.

It may help to mentally start each implementation comment with something like "this is interesting/surprising because ...".

In this case we're taking pains to ensure that the new app has the same name and options as the prior one so that the persistent storage is reused from the previous instance. It can be helpful to contrast against the "obvious" way of doing things (i.e. just calling `[self firestore]` intentionally produces a separate storage directory which defeats the test).

With that said, there are circumstances in which a particularly gnarly block of code may need a description of what it's doing but this is uncomplicated except for our motivations for doing it that way. The comment captures our motivation because that's not obvious from the code.

Reply:
Thanks for taking the time to type this out -- it really helps me better understand how to approach comments in a systematically useful way.

> Again, this comment describes _what_ the code is doing without any insight into _why_ it's doing so. Your intuition to make a comment here is right, you just need to comment on why--the code already says what it's doing.
> 
> It may help to mentally start each implementation comment with something like "this is interesting/surprising because ...".
> 
> In this case we're taking pains to ensure that the new app has the same name and options as the prior one so that the persistent storage is reused from the previous instance. It can be helpful to contrast against the "obvious" way of doing things (i.e. just calling `[self firestore]` intentionally produces a separate storage directory which defeats the test).
> 
> With that said, there are circumstances in which a particularly gnarly block of code may need a description of what it's doing but this is uncomplicated except for our motivations for doing it that way. The comment captures our motivation because that's not obvious from the code.



> It may help to mentally start each implementation comment with something like "this is interesting/surprising because ...".

Thanks for taking the time to type this out -- this gives me a systemically useful way to create useful code comments.